### ISSUE TITLE
Contrived parameters in samtools depth? 

### ISSUE THREAD

**Issue by dariober** [samtools/samtools#1702]
I guess there is good reason for this that I'm not seeing... `samtools depth` (v1.15.1) has options -g/-G:

```
samtools depth
Usage: samtools depth [options] in.bam [in.bam ...]

Options:
...
  -g INT       Remove specified flags from default flag filter
  -G INT       Add specified flags to the default flag filter
...
``` 

these seem a bit contrived. Why not simply have options `--incl-flags/--excl-flags` like mpileup and view?

Once I'm here... option `-H` is repeated twice:

```
samtools depth
Usage: samtools depth [options] in.bam [in.bam ...]

Options:
  -a           Output all positions (including zero depth)
  -a -a, -aa   Output absolutely all positions, including unused ref seqs
  -r REG       Specify a region in chr or chr:from-to syntax
  -b FILE      Use bed FILE for list of regions
  -f FILE      Specify list of input BAM/SAM/CRAM filenames
  -X           Use custom index files (in -X *.bam *.bam.bai order)
  -g INT       Remove specified flags from default flag filter
  -G INT       Add specified flags to the default flag filter
  -H           Print a file header line
  -l INT       Minimum read length [0]
  -o FILE      Write output to FILE [stdout]
  -q, --min-BQ INT
               Filter bases with base quality smaller than INT [0]
  -Q, --min-MQ INT
               Filter alignments with mapping quality smaller than INT [0]
  -H           Print a file header
  -J           Include reads with deletions in depth computation
  -s           Do not count overlapping reads within a template
      --input-fmt-option OPT[=VAL]
               Specify a single input file format option in the form
               of OPTION or OPTION=VALUE
  -@, --threads INT
               Number of additional threads to use [0]
      --verbosity INT
               Set level of verbosity
```

**Comment by whitwham** [comment]
>  Why not simply have options --incl-flags/--excl-flags like mpileup and view?

Mostly because no one has gotten around to adding them. These programs have been written and amended by multiple people, from different groups, over 14 years.  Not all the developments have been consistent.  While we try to make the samtools commands more consistent than they were there is a long way to go.  Also, while we can add long options to `depth`, we cannot remove the old ones because we might break the pipeline of someone who uses them.

> Once I'm here... option -H is repeated twice

Well that is just a bug.  It looks like someone was reordering the options into alphabetical order and forgot to delete the second -H.

**Comment by jkbonfield** [comment]
The depth options were added in 3b0753f8 (first releases as 1.10).   Samtools depth had a total rewrite in 1.13, but we kept the same options for compatibility so for the purposes of this issue that's irrelevant.

https://github.com/samtools/samtools/pull/1442#issuecomment-871244137 is key for the -g/-G issue.  Confusingly `depth -G` serves the same role as `view -F`, with `view -G` doing something completely different.  Adding long options however give us the chance of sanitizing things so I'll look at that.

There is also a subtle difference in how `--excl-flags` works.  `mpileup --excl-flags` *sets* the filter bit pattern, thus it overrides the default of "UNMAP,SECONDARY,QCFAIL,DUP".  `view --excl-flags` *or*s the filter bit pattern, but it starts with a value of zero so it amounts to a similar action.  `depth -G` has the same *or* action as view, but it starts with a value of `BAM_FUNMAP | BAM_FSECONDARY | BAM_FDUP | BAM_FQCFAIL`, so it needs a mechanism or striking out predefined settings, hence `-g`.  If it did the same *set* strategy as mpileup then it'd be OK, but it doesn't.  Note `-g` is therefore not the same as `--incl-flags`.  It doesn't set a required-flags list and `depth` has no equivalent functionality for that and no-one has requested it.

In short: it's a total mess!  We wouldn't start from here if we could fix it.  Neither can we simply rationalise it as that breaks existing tools.  With hindsight we should have noticed this incompatibility when preparing for 1.10 so the initial -G option to depth was saner, but no one did.
